### PR TITLE
[MPS] Migrate scatter/gather ops to Metal with async bounds checking

### PR DESCRIPTION
**Pull request by malfet**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #184028

Migrates family of `gather` / `scatter` ops (including reduce parts) from from the legacy MPSGraph path to native Metal kernels, registers MPS into the existing stub patterns shared with CPU/CUDA.

- Uses standard dense/strided/broadcast(or scalar) pattern to match/outperform MPSGraph generated kernels.
- Uses `AtomicType<T>::atomic_add` for add flavor , and prod/amin/amax rely on `atomic_binary_op` template (packed-CAS loop).
- Bounds checks live inline in every kernel and report to the stream's async error buffer.
- Deleted `view_as_real` workarounds for complex types.

## Signed int64 amin/amax

Apple's `<metal_atomic>` exposes `atomic_min_explicit` / `atomic_max_explicit` on `device ulong*` (unsigned only) starting on macOS 15 GPUs. For signed int64 we bracket the scatter with an XOR-sign-bit pass (encode → atomic_min/max on the ulong view → decode), so signed ordering maps to the unsigned hardware op. Gated at runtime on macOS 15+.

`prod` on int64 isn't reachable through this trick (Metal has no `atomic_compare_exchange` or `atomic_prod` on 64-bit, and prod isn't a monotone reduction), so it stays in the skip list.

## Large tensor support

`[[thread_position_in_grid]]` is at most a `uint`, so the host chunks dispatch into <=UINT_MAX-thread launches and each thread reads a `tid_offset` constant. Scatter on tensors with index numel > 2^32 works transparently.

## Bool reduce ops

Added `atomic_binary_op` to `AtomicType<bool>` in `c10/metal/atomic.h` (one-liner delegating to the existing packed-uint CAS helper), so bool prod/amin (AND) and amax (OR) work through the same machinery as other dtypes. `scatter_reduce(mean, bool)` is rejected with `NotImplementedError` to match CPU.

## Performance

vs PyTorch 2.12.0 (median delta across {(1024,1024) dim=1, (512,4096) dim=1, (4096,512) dim=0, (128,128,128) dim=2}; negative = new Metal path is that much faster than the old MPSGraph baseline):

| Op | bool | int8 | uint8 | int16 | int32 | int64 | f16 | bf16 | f32 | cfloat |
|---|---|---|---|---|---|---|---|---|---|---|
| `scatter_` set (tensor src) | -51% | -24% | -49% | -28% | -29% | -37% | -28% | -23% | -30% | -62% |
| `scatter_` set (scalar) | -50% | -31% | -50% | -36% | -38% | -52% | -39% | -34% | -40% | -70% |
| `scatter_add_` | -48% | -39% | -38% | -45% | -40% | -58% | -36% | -40% | -32% | -57% |
| `scatter_reduce_(amax)` | -45% | -42% | -42% | -45% | -33% | NEW¹ | -45% | -47% | -34% | n/a |
| `scatter_reduce_(prod)` | -46% | -42% | -43% | -45% | -32% | n/a² | -40% | -44% | -34% | n/a |

¹ int64 amin/amax was previously unsupported on MPS (skipped at OpInfo). Now works on macOS 15+ via the sign-flip + ulong atomic_min/max bracket. 92-273 us depending on shape.
² int64 prod still unsupported (no Metal 64-bit atomic_binary_op even with the encoding trick).

`one_hot` itself (the original issue) gets 30-60% faster across shapes since it exercises the dense scalar path directly.

No regression on any tested shape × dtype × op combination.

<details>
<summary>Benchmark script</summary>

```python
import time, torch

def bench(fn, iters=100, warmup=20):
    for _ in range(warmup): fn()
    torch.mps.synchronize()
    t0 = time.perf_counter()
    for _ in range(iters): fn()
    torch.mps.synchronize()
    return (time.perf_counter() - t0) * 1e6 / iters

DTYPES = [torch.bool, torch.int8, torch.uint8, torch.int16, torch.int32, torch.int64,
          torch.float16, torch.bfloat16, torch.float32, torch.complex64]
SHAPES = [((1024, 1024), 1), ((512, 4096), 1), ((4096, 512), 0), ((128, 128, 128), 2)]

def make(dt, shape, dim):
    self = torch.zeros(shape, dtype=dt, device="mps")
    if dt == torch.bool:           src = torch.ones(shape, dtype=dt, device="mps")
    elif dt.is_floating_point or dt.is_complex: src = torch.randn(shape, device="mps").to(dt)
    else:                          src = torch.randint(0, 10, shape, device="mps").to(dt)
    idx = torch.randint(0, shape[dim], shape, device="mps", dtype=torch.long)
    return self, idx, src

for shape, dim in SHAPES:
    for dt in DTYPES:
        s, i, v = make(dt, shape, dim)
        scalar = True if dt == torch.bool else (1.5 if dt.is_floating_point or dt.is_complex else 1)
        for name, fn in [
            ("set/tensor", lambda: s.scatter_(dim, i, v)),
            ("set/scalar", lambda: s.scatter_(dim, i, scalar)),
            ("add",        lambda: s.scatter_add_(dim, i, v)),
            ("amax",       lambda: s.scatter_reduce_(dim, i, v, reduce="amax", include_self=True)),
            ("prod",       lambda: s.scatter_reduce_(dim, i, v, reduce="prod", include_self=True)),
        ]:
            try: us = bench(fn)
            except Exception: us = float("nan")
            print(f"{name:12} {str(dt):<20} {str(shape):<22} {us:>8.2f}")
```

</details>

Fixes https://github.com/pytorch/pytorch/issues/170507

( as `F.one_hot` is composite -- it lowers to `zeros() + scatter_(-1, self.unsqueeze(-1), 1)`)

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: aten/src/ATen/native/mps/kernels/Scatter.metal [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    constant long* src_strides [[buffer(5)]],
+    constant long* index_strides [[buffer(6)]],
+    constant uint3& ndim_dim [[buffer(7)]],
+    constant long& dim_size [[buffer(8)]],
+    device ErrorMessages* error_buf [[buffer(9)]],
+    uint thread_index [[thread_position_in_grid]]) {
+  const uint ndim = ndim_dim.x;
+  const uint dim = ndim_dim.y;
+
+  long pos[max_ndim];
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/kernels/Scatter.metal]: Better to not use metal array?

**Review by Skylion007 — APPROVED** [review]

**Comment by malfet** [comment]
@claude review this

File: aten/src/ATen/native/mps/operations/ScatterGather.mm [hunk 2]
```diff
@@ -15,401 +16,479 @@
...
-                                const Tensor& src,
-                                const Tensor& output,
-                                std::string func_name,
-                                const std::string_view reduce) {
-  using namespace mps;
+REGISTER_DISPATCH(gather_stub, &gather_mps_kernel);
 
-  if (self_arg.numel() == 0 || index.numel() == 0 || src.numel() == 0) {
-    return;
+static const char* reduce_op_to_mps_string(const ReductionType& op) {
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/ScatterGather.mm]: Could return std::string_view I guess

Comment (Skylion007) [reply aten/src/ATen/native/mps/operations/ScatterGather.mm]: Then you don't need to cast as much later

File: aten/src/ATen/native/mps/operations/ScatterGather.mm [hunk 3]
```diff
@@ -15,401 +16,479 @@
...
-    auto feeds = dictionaryFromPlaceholders(selfPlaceholder, srcPlaceholder, indexPlaceholder);
-    runMPSGraph(getCurrentMPSStream(), cachedGraph->graph(), feeds, outputPlaceholder);
+  TORCH_CHECK(self.scalar_type() == src.scalar_type(), "scatter(): self and src must have the same scalar type");
+  // Complex types map to float2/half2 in the Metal kernels (see
+  // scalarToMetalTypeString), so no view_as_real shim is needed.
+  auto self_view = self.dim() == 0 ? self.view({1}) : self;
+  auto src_view = maybe_expand_0_dim(src);
+  auto index_view = maybe_expand_0_dim(index);
+  dim = at::maybe_wrap_dim(dim, self_view.dim());
+  TORCH_CHECK(dim >= 0 && dim < self_view.dim(), "scatter(): Indexing dim ", dim, " is out of bounds of tensor");
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/ScatterGather.mm:405]: These should be TORCH_CHECK_VALUE unless it's already broke on other tests / impls sigh

Comment (malfet) [reply aten/src/ATen/native/mps/operations/ScatterGather.mm:405]: Let me(Claude) submit a followups for all in-core backends
